### ISSUE TITLE
Missing support for zpub/ypub/... public key formats

### ISSUE THREAD

**Issue by jviki** [rust-bitcoin/rust-bitcoin#2190]
Are there any plans to include support for pub/priv key formats other than xpub?

I tried to obtain zpub key via trezor-client (https://github.com/trezor/trezor-firmware/tree/main/rust/trezor-client) and received Bip32(UnknownVersion([4, 178, 71, 70])) which puzzled me for a while until I discovered that rust-bitcoin only supports xpubs...

**Comment by apoelstra** [comment]
No, we are never going to support these. I'm not sure what crate will support them.

If you want your secret keys to be used by specific addresses you should be using output descriptors, which are far more general and provided by the rust-miniscript crate.

**Comment by jviki** [comment]
Thanks for clarification. I don't care too much, it's just that trezor-client is based on rust-bitcoin and thus it does not work as expected. I don't think I can choose a crate here...

**Comment by apoelstra** [comment]
I think you'll need to file an issue on trezor-client -- since this is a trezor-specific thing (and a few other hardware wallets) I think it's on them to support them.

Even if rust-bitcoin were to add ypub and zpub types, they wouldn't resemble xpubs at all (this is one reason they're on the "won't implement" list for this library) and instead would resemble descriptors, so trezor-client would still need to do a bunch of integration work.

**Comment by junderw** [comment]
I think restricting the bip32 types to only Bitcoin is not technically BIP32 compliant...

Maybe we should rename the module to `hd_keys` or something.

Technically zpub and whatnot are concepts that can and do exist within the parameters of BIP32, so the fact that it's a hard error to decode and we have no way (like a trait or something) that an external crate could extend it to include other valid BIP32 schemes tells me that maybe we should rename it.

Or offer some way for people to extend it to suit their needs.

I would be willing to work on that if you're against renaming it.

**Comment by junderw** [comment]
As a reference: in bitcoinjs-lib our Network concept is an interface, not a concrete type, so people can implement their own zpub network etc. which will allow for encoding and decoding of zpubs.

**Comment by Kixunil** [comment]
@junderw I don't think that's the case. IIUC ypub/zpub just define specific numbers in the path. So if you write your own ypub/zpub parser that decodes the key, then instantiate it and use appropriate derivation it should work.

**Comment by junderw** [comment]
Yes, you can basically copy paste the `decode` function and remove the network matching and just slap a `Network::Bitcoin` in there regardless, and just check if the version bytes match whatever the version bytes for zpub etc are.

But my point still stands.

BIP32 allows for more than 2 versions (mainnet and testnets) and yet the bip32 module does not. Therefore there is a huge possibility space of valid BIP32 (ie. Litecoin HD keys) that can not be represented by the bip32 module's types due to the tight coupling with the `Network` type which only includes Bitcoin related networks.

**Comment by junderw** [comment]
I am leaning towards a rename, since de-coupling is a big commitment once everything is so tightly coupled.

We can mention BIP32 in the doc comments of the newly named module, and that would help SEO / people searching the docs.

**Comment by Kixunil** [comment]
Network matching looks orthogonal to zpub/ypub but yes, we're not adhering to spec and I think we should (even though I really dislike complicating the API).

```rust
pub enum Bip32Network {
    Bitcoin(Network),
    Unknown(u32), // or whatever the type is, didn't check
}

pub struct Xpub {
    pub network: Bip32Network,
// ...
}
```

And yes, I want to separate BIP32 and you have a point that it's not an entirely bitcoin thing. (Although "B" in BIP...) So something like `hd_derivation` would still make sense.

> I am leaning towards a rename

I do as well for a different reason: easier for people who don't know BIPs by heart (although this one is very famous and well known).

**Comment by junderw** [comment]
@jviki BTW you could do something like this, but it would be hard coded to Bitcoin mainnet and the `encode` method will return an xpub. (I just basically copy pasted the code from `decode` and the `FromStr` impl)

```rust
/// The version bytes for zpub extended keys
pub const ZPUB_VERSION: [u8; 4] = [4, 178, 71, 70];

/// Used for decoding custom xpub (zpub etc.)
pub fn hd_pub_from_custom_str(s: &str, version: [u8; 4]) -> Result<Xpub, Error> {
    let data = base58::decode_check(s)?;

    if data.len() != 78 {
        return Err(Error::WrongExtendedKeyLength(data.len()));
    }

    if !data.starts_with(&version) {
        let (b0, b1, b2, b3) = (data[0], data[1], data[2], data[3]);
        return Err(Error::UnknownVersion([b0, b1, b2, b3]));
    }

    Ok(Xpub {
        network: Network::Bitcoin,
        depth: data[4],
        parent_fingerprint: data[5..9]
            .try_into()
            .expect("9 - 5 == 4, which is the Fingerprint length"),
        child_number: u32::from_be_bytes(data[9..13].try_into().expect("4 byte slice")).into(),
        chain_code: data[13..45]
            .try_into()
            .expect("45 - 13 == 32, which is the ChainCode length"),
        public_key: secp256k1::PublicKey::from_slice(&data[45..78])?,
    })
}
```

**Comment by apoelstra** [comment]
@junderw can you provide a citation in BIP32 suggesting that we should support networks beyond bitcoin and testnet?

If we are out of compliance certainly we should fix that. But as @Kixunil says, none of this has anything to do with ypubs or zpubs, which do not appear in BIP32.

**Comment by tcharding** [comment]
Slightly off topic but seems like an opportune time/place to discuss pulling `bip32` out of `rust-bitcoin`:

- I've started working on pulling PSBT code out of `rust-bitcoin` (and `rust-miniscript`) into a separate crate
- PSBT and BIP-32 are both wallet things
- We have a repo on `github.com/rust-bitcoin/rust-wallet`

Why don't we pull `bip32` and combine it with my work on PSBT? Seems to solve a bunch of problems at once.

**Comment by tcharding** [comment]
Also its way more fun at times to hack like mad on separate repos since we get really bogged down with reviews here in `rust-bitcoin`.

**Comment by junderw** [comment]
> we should support networks beyond bitcoin and testnet?

@apoelstra There is no concept of a network in BIP32. The version bytes are just 4 bytes. The word "Bitcoin" is only used to reference how Bitcoin wallets work "currently" and / or draw comparisons to base58 addresses for P2PKH. The word "network" is not used anywhere, and the words "mainnet" and "testnet" are used to give the version numbers for those respective networks, but nowhere is it mentioned explicitly that these are the only 4 valid version numbers.

In November 2020, 8 years after wide adoption, someone added Test Case 5 which includes "invalid extended keys" that reject an "unknown extended key version"... but this is the only place in BIP32 that even implicitly hints that the 4 byte version can not make use of the full 32 bit possibility space.

**Comment by apoelstra** [comment]
> Why don't we pull `bip32` and combine it with my work on PSBT? Seems to solve a bunch of problems at once.

Because it's one tiny module, basically, which hasn't changed (except to tighten up some types) in years and which has never gotten any complaints until this issue.

>In November 2020, 8 years after wide adoption, someone added Test Case 5 which includes "invalid extended keys" that reject an "unknown extended key version"... but this is the only place in BIP32 that even implicitly hints that the 4 byte version can not make use of the full 32 bit possibility space.

I think, explicitly listing four permissible values for the version "implicitly" says that those are the only four values. And the test vector clinches it.

But sure, I'm happy to support arbitrary version bytes. I don't like restrictions of this form (and don't like them with addresses either; but since we were able to drop the HRP restriction for segwit addresses I care less about legacy ones).

**Comment by junderw** [comment]
https://github.com/bitcoin/bips/pull/921

To be fair....... even I overlooked this subtle breaking change. (I'm the first response to the PR that introduced the test case.

**Comment by Kixunil** [comment]
> which has never gotten any complaints

I believe I did complain that the API is really not great and if I didn't I do now. Many things, don't want to spend time listing them right now. Also I don't understand why applications like electrs should spend time compiling these things. I want it separated and changed.

**Comment by junderw** [comment]
I agree that it should be separated into a separate crate, with ties to rust-bitcoin removed so that dependent crates don't need to pull in rust-bitcoin.
